lnwallet/channel.go

```diff
@@ -3854,39 +3986,46 @@ func (lc *LightningChannel) validateCommitmentSanity(theirLogCounter,
 	if err != nil {
 		return err
 	}
-	feePerKw := filteredView.feePerKw
 
-	// Calculate the commitment fee, and subtract it from the initiator's
-	// balance.
-	commitFee := feePerKw.FeeForWeight(commitWeight)
-	commitFeeMsat := lnwire.NewMSatFromSatoshis(commitFee)
-	if lc.channelState.IsInitiator {
-		ourBalance -= commitFeeMsat
-	} else {
-		theirBalance -= commitFeeMsat
+	// We need to first check whether we went negativ. We might underflow in
+	// computeView because lnwire.MilliSathoshi is an unsigned type.
+	if int64(ourBalance) < 0 {
+		return ErrBelowChanReserve
 	}
-
-	// As a quick sanity check, we'll ensure that if we interpret the
-	// balances as signed integers, they haven't dipped down below zero. If
-	// they have, then this indicates that a party doesn't have sufficient
-	// balance to satisfy the final evaluated HTLC's.
-	switch {
-	case int64(ourBalance) < 0:
-		return fmt.Errorf("%w: negative local balance",
-			ErrBelowChanReserve)
-
-	case int64(theirBalance) < 0:
-		return fmt.Errorf("%w: negative remote balance",
-			ErrBelowChanReserve)
+	if int64(theirBalance) < 0 {
+		return ErrBelowChanReserve
 	}
 
+	feePerKw := filteredView.feePerKw
+
 	// Ensure that the fee being applied is enough to be relayed across the
 	// network in a reasonable time frame.
 	if feePerKw < chainfee.FeePerKwFloor {
 		return fmt.Errorf("commitment fee per kw %v below fee floor %v",
 			feePerKw, chainfee.FeePerKwFloor)
 	}
 
+	// The channel opener has to account for the commitment fee. This
+	// includes also a buffer type. Depending on whether we are the opener
+	// of the channel we either want to enforce a buffer on the local
+	// amount.
+	var bufferAmt lnwire.MilliSatoshi
+	if lc.channelState.IsInitiator {
+		ourBalance, bufferAmt, err = lc.applyCommitFee(
+			ourBalance, commitWeight, feePerKw, true, buffer)
+		if err != nil {
+			return err
+		}
+	} else {
+		theirBalance, bufferAmt, err = lc.applyCommitFee(
```

Comment:
So if we are not the channel funder we won't check their balance to account for the fee buffer? Why dont we also check it?

Reply:
That's a very good idea 👌, this will help to prevent this bug when only one side has been updated to the newest lnd release, nice !

Reply:
Thought about this a second time and I think we cannot do this. When we also enforce a feebuffer when sending but the other node does not want to enforce the feebuffer (still on older software) we can end up in a scenario that the peer depletes his side even more and now the channel is stuck because he cannot rebalance into the channel because we will not put any HTLC on it. So we should not intervene when we are NOT the opener I think.